add `sync::OnceExt` and `sync::OnceLockExt` traits

davidhewitt opened the pull request:
Related to #4513, #4671 

It turns out that within our test suite the racy behaviour of `GILOnceCell` is already problematic on GIL-enabled builds, and is just waiting for additional GIL switches to bait out problems 😬. In particular in `test_declarative_module` we have a race which might attempt to create a module twice; we explicitly forbid this because we don't support subinterpreters.

This PR introduces `pyo3::sync::OnceExt`, which adds helper methods to `std::sync::Once` which avoid deadlocking with the GIL by releasing if necessary before blocking.

This solves the issue with `test_declarative_module` by using the `Once` to guarantee we only ever create the module once.

I think we should document this in the FAQ and freethreading guides, and probably this is good justification for getting #4513 and maybe also some `OnceCellExt` traits into 0.23. cc @ngoldbaum, I guess this would delay the release by a few days.

ngoldbaum commented:
> I guess this would delay the release by a few days

That's OK. Single-initialization comes up *a lot* so we should take our time and get it right.

I also want to add a section to the guide about how to use threads from both Python and rust and how free-threading changes things. I may expand the existing parallelism docs if needed.

ngoldbaum commented on src/sync.rs:
```diff
@@ -473,6 +480,58 @@ where
+            // Some other thread filled this Once, so we need to restore the GIL state.
+            unsafe { ffi::PyEval_RestoreThread(ts) };
+        }
+    }
```
I wrote a test for this trait based on the standard library example for `try_once_force` and it crashed Python because the thread state was corrupted. I think this code won't restore the thread state correctly if a panic happens inside `f()` and the fix is to use the RAII guard pattern. Assuming that's right, I'll go ahead and push a fix along with the test.

mejrs replied:
Does the guard fix your problem? That'd be surprising to me, because I reviewed the previous implementation and believed it was sound.

mejrs replied:
Oh nevermind; I read the Once documentation and if it is poisoned it will panic before the closure is called.

ngoldbaum commented:
I looked at including `OnceLockExt` in this PR as well and I got a fair bit of the way through that (see https://github.com/ngoldbaum/pyo3/commit/f0eb7e07f2aa3598e6f18a8a26703fd32d9a1df3, feel free to pull that and use it if you want) but realized we don't have a way to conditionally compile based on the rust compiler version, so maybe we need to bump the MSRV? Or is there a way to account for the fact that the MSRV doesn't have OnceLock?

mejrs commented:
> but realized we don't have a way to conditionally compile based on the rust compiler version

We do? https://github.com/PyO3/pyo3/blob/main/pyo3-build-config/src/lib.rs#L138

ngoldbaum commented:
Sorry, missed that! Thanks for the link.

ngoldbaum commented:
> We do?

I *think* I did it right, but clippy complains that I'm using features that aren't in the MSRV. OK to ignore that or did I mess something up?

ngoldbaum commented on src/sync.rs:
```diff
@@ -589,4 +725,54 @@ mod tests {
+            });
+        });
+        assert_eq!(cell.get(), Some(&12345));
+    }
```
suggestions welcome for a more interesting test

ngoldbaum commented on guide/src/free-threading.md:
````diff
@@ -152,6 +152,60 @@ We plan to allow user-selectable semantics for mutable pyclass definitions in
+});
+# }
+```
+
````
I probably want to rewrite this example to use OnceLock

mejrs commented:
> > We do?
> 
> I _think_ I did it right, but clippy complains that I'm using features that aren't in the MSRV. OK to ignore that or did I mess something up?

That's annoying. You can set it to allow here: https://github.com/PyO3/pyo3/blob/main/Cargo.toml#L153

ngoldbaum commented:
I just locally ignored it, I still think that clippy lint is useful and it doesn't look like we use features that aren't in the MSRV very often.

davidhewitt commented:
Thanks, this looks great to me! I simplified the `Guard` type a little (think the option no longer needed), and corrected a couple docs typos. Will merge!

ngoldbaum commented:
The new tests that use threads need to be skipped on emscripten.
